ext/spl/spl_fixedarray.c

```diff
@@ -609,17 +609,12 @@ PHP_METHOD(SplFixedArray, __serialize)
 	}
 
 	/* members */
-	if (intern->std.properties) {
-		ZEND_HASH_FOREACH_STR_KEY_VAL(intern->std.properties, key, current) {
-			/* The properties hash table can also contain the array elements if the properties table was already rebuilt.
-			 * In this case we'd have a NULL key. We can't simply use the properties table in all cases because it's
-			 * potentially out of sync (missing elements, or containing removed elements) and might need a rebuild. */
-			if (key != NULL) {
-				zend_hash_add_new(Z_ARRVAL_P(return_value), key, current);
-				Z_TRY_ADDREF_P(current);
-			}
-		} ZEND_HASH_FOREACH_END();
-	}
+	ZEND_HASH_FOREACH_STR_KEY_VAL_IND(ht, key, current) {
+		if (key != NULL) {
```

Comment:
Can you please describe me when the key is null? I wasn't be able to understand it 100% based on the comment above

Reply:
I see I removed the comment, I need to place that back.

In any case: if var_dump (or something else that rebuilds the properties table) has been called before serialize(), then the properties table will include entries for the array elements. In those cases the key == NULL, and h == the index of the element.

For example:
```php
<?php
$test = new SplFixedArray(2);
var_dump($test);
var_dump(serialize($test)); // If you add a printf to the foreach loop you'll see that the entries for array index 0 and 1 are also printed. But this does not happen if you remove the preceding var_dump, because then the properties table hasn't been rebuilt yet.
```

Reply:
Also the difference between this version of the patch and the original is the following:
In the original patch I only added support for the INDIRECT zvals by appending _IND.
Shortly after merging I noticed a problem, so I reverted the merge and re-opened this PR.
The problem I noticed is that the code was using the wrong HashTable, their entries could still be stale (see the most recent commit which adds a testcase for this).

Reply:
Thanks for the explanation! The new code comment is very helpful!